Add a nice header to our README.md

Urgau opened the pull request:
This PR improves our README, it is greatly inspired by [esbuild](https://github.com/evanw/esbuild/blob/9d1777f23d9b64c186345223d92f319e59388d8b/README.md) README.

Context: As was reading https://johnjago.com/great-docs/ I though we could greatly improve ours. So that's what I did.

It provides direct "quick-access" links to pages in rust-lang.org.

The "Why Rust?" section is ~~a direct copy/paste of the same~~ modified version of section in [rust-lang.org](https://www.rust-lang.org/).

| Before | After |
|--------|-------|
| ![before](https://github.com/rust-lang/rust/assets/3616612/4afb6753-18a9-4881-919e-2a79328beb5a) | ![after](https://github.com/rust-lang/rust/assets/3616612/408dea7b-5d97-4bb8-a77e-35541ddd50cb) <details> ![after](https://github.com/rust-lang/rust/assets/3616612/76e4a8c6-b61c-46e8-b5ba-4f09c13cfc94) </details> <details> ![after -1](https://github.com/rust-lang/rust/assets/3616612/ed50675c-8301-457c-8b9a-a1199c515fb7) </details> |

Note: I removed the manual TOC, since GitHub provides it's own at the top right corner and I don't think it's needed anymore.
Same for the notice about the readme being for users, it's now clear enough and that notice was distracting anyway.

Urgau committed:
Add nice header to our README.md

Nilstrieb commented on README.md:
```diff
@@ -1,27 +1,33 @@
-
-[![Rust Community](https://img.shields.io/badge/Rust_Community%20-Join_us-brightgreen?style=plastic&logo=rust)](https://www.rust-lang.org/community)
+<p align="center">
+  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software" src="https://www.rust-lang.org/static/images/rust-social-wide.jpg">
```
imo this image is way too big, it's not very useful and just takes away space.

Urgau replied:
yeah, the social image from rust-lang.org has unfortunately the right ratio to be very big; I will try to reduce it's size

Urgau replied:
done.

Nilstrieb commented on README.md:
```diff
@@ -1,27 +1,33 @@
-**Note: this README is for _users_ rather than _contributors_.**
-If you wish to _contribute_ to the compiler, you should read
-[CONTRIBUTING.md](CONTRIBUTING.md) instead.
+## Why Rust?
```
i'm also not sure whether it's worth having this. i don't expect that people visiting this repository need to be convinced that rust is good, i would expect that people coming to the repository are rust users already. maybe i'm wrong though.

Urgau replied:
maybe, maybe not; I would expect to have at least some people who don't know Rust at all, to come to our README first before anything else, and giving them (without indirection) reason to like the language is IMO only positive

workingjubilee replied:
I don't disagree with _having_ the Why but we can keep it very terse and make fewer unsubstantiated / overly-subjective claims. Some people I've met claim Rust _doesn't_ have great documentation, because they are used to ecosystems which are even more thoroughly documented (though I disagree about "great"). However we do have e.g. integrated doc tooling.

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,33 @@
+
+### Performance
+
+Rust is blazingly fast and memory-efficient: with no runtime or garbage collector, it can power performance-critical services, run on embedded devices, and easily integrate with other languages.
```
If Rust doesn't have a runtime, how do you explain the runtime startup?

https://github.com/rust-lang/rust/blob/master/library/rtstartup/rsbegin.rs

We both know what people mean, but this is the surface of the documentation for the language, it needn't be written in a marketing-speak tone.

Urgau replied:
There is no longer any mention of "runtime".

Urgau commented on README.md:
```diff
@@ -1,27 +1,33 @@
-**Note: this README is for _users_ rather than _contributors_.**
-If you wish to _contribute_ to the compiler, you should read
-[CONTRIBUTING.md](CONTRIBUTING.md) instead.
+## Why Rust?
```
@workingjubilee in https://github.com/rust-lang/rust/pull/123093#discussion_r1539751987

> I don't disagree with _having_ the Why but we can keep it very terse and make fewer unsubstantiated / overly-subjective claims. Some people I've met claim Rust _doesn't_ have great documentation, because they are used to ecosystems which are even more thoroughly documented (though I disagree about "great" with them, they are allowed to have their own criteria). However we do have e.g. integrated doc tooling.

Urgau replied:
These statements were taken literally from the website, so let's edit them at source first. Otherwise, it will be a bit of a disavowal of them.

workingjubilee replied:
I think it's fine if those are written in a different tone because it's not on the primary README for the distribution of the program which should first and foremost serve those who need information about this motley bundle of files.

Urgau replied:
Feel free to suggest modifications/improvements.

<details>
<summary>LLM improved version</summary>

### Performance
Rust is known for its high speed and efficient memory usage, making it suitable for performance-intensive tasks such as powering critical services, running on embedded devices, and seamless integration with other programming languages.

### Reliability
Rust's robust type system and ownership model ensure memory and thread safety, reducing the likelihood of bugs at compile-time.

### Productivity
Rust offers comprehensive documentation, a user-friendly compiler with helpful error messages, and advanced tooling including an integrated package manager, build tool, multi-editor support with auto-completion and type inspections, an auto-formatter, and additional features.

</details>

workingjubilee replied:
Yeah hitting our points hard and fast seems like an improvement on the previous version, hmm.

Mark-Simulacrum commented:
r? @workingjubilee 

(Feel free to re-roll further. I don't have any strong opinion on this kind of change.)

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,25 @@
+- **Reliability:** Our rich type system and ownership model ensure memory and thread safety, reducing bugs at compile-time.
 
-</details>
+- **Productivity:** Comprehensive documentation, user-friendly compiler, and advanced tooling including package manager, build tool, auto-formatter, and editor support.
```
This deviation from the website's way of expressing it is again driven by "this is the README for the compiler's source repo, so we should use it to introduce people to what they can find in this repo or related repos".

Urgau replied:
> I feel like our compiler is slightly more-precise than "user-friendly": we want good diagnostics, and we treat bad diagnostics as bugs if they are at all improvable.

Completely agree, but I don't know how to say that with only 2/3 words. I therefor changed it to "user-focused" instead of "user-friendly" to emphasise the fact that we care about the users.

> I also think we can go ahead and cross-link here: we can identify that we ship this stuff via Rustup, and we can name cargo, rustfmt, rust-analyzer, and what they are, with hyperlinks.

Also agree, I've added a hyperlink to all the major tools we ship.

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,25 @@
+<p align="center">
+  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software" src="https://www.rust-lang.org/static/images/rust-social-wide.jpg" width="512" height="256">
+  <br>
+  <a href="https://www.rust-lang.org/">Website</a> |
+  <a href="https://www.rust-lang.org/learn/get-started">Getting started</a> |
+  <a href="https://www.rust-lang.org/learn">Learn</a> |
+  <a href="https://www.rust-lang.org/learn#learn-use">Documentation</a> |
+  <a href="CONTRIBUTING.md">Contributing</a>
```
These can continue to be normal Markdown links and thus all as one line, inline, with the links provided in Markdown "footer" later, making it clear how it should look. As long as you use line breaks appropriately between things, you can intermingle HTML and Markdown syntax in a CommonMark or GitHub-Flavored Markdown document (because GHF strives to be compatible with CommonMark). So just use a line break before and after the Markdown syntax. The HTML will still be included directly and will still affect the Markdown after.

Urgau replied:
Done.

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,25 @@
-[![Rust Community](https://img.shields.io/badge/Rust_Community%20-Join_us-brightgreen?style=plastic&logo=rust)](https://www.rust-lang.org/community)
+<p align="center">
+  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software" src="https://www.rust-lang.org/static/images/rust-social-wide.jpg" width="512" height="256">
+  <br>
```
Per my last comment you probably don't even need this, just a space.

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,25 @@
-# The Rust Programming Language
-
-[![Rust Community](https://img.shields.io/badge/Rust_Community%20-Join_us-brightgreen?style=plastic&logo=rust)](https://www.rust-lang.org/community)
+<p align="center">
```
If this all looks the same with either the `header` tag or the `section` tag those would be preferable. Also the `align` attribute is deprecated, so assuming it renders the same please write this as
```suggestion
<header style="text-align:center;">
```

Urgau replied:
This is sadly not possible, as GitHub removes any inline style.

We also can't use `<header>` or `<section>` since neither of them supports the `align="center"` hack.

And we can't continue to use `<p>` since it doesn't work with real Markdown links, so I changed it to `<div align="center">`.

workingjubilee replied:
I "love" this discovery. Apologies, I tried it out in a CommonMark demo before recommending it and didn't realize the [GFM spec](https://github.github.com/gfm/) was out of date.

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,25 @@
+  <a href="https://www.rust-lang.org/learn">Learn</a> |
+  <a href="https://www.rust-lang.org/learn#learn-use">Documentation</a> |
+  <a href="CONTRIBUTING.md">Contributing</a>
+</p>
```
```suggestion
</header>
```

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,25 @@
-- [Getting Help](#getting-help)
-- [Contributing](#contributing)
-- [License](#license)
-- [Trademark](#trademark)
+- **Reliability:** Our rich type system and ownership model ensure memory and thread safety, reducing bugs at compile-time.
 
```
I don't think these spaces between bullets are necessary.

Urgau replied:
They change the readability of the bullet points. Without spaces I think it's too close together.

Without spaces:
![image](https://github.com/rust-lang/rust/assets/3616612/b2c033b5-dece-486e-8afd-ecacf5c03b3d)

With spaces:
![image](https://github.com/rust-lang/rust/assets/3616612/1a93a5ac-f7f1-4d18-b5cd-230e7ee44778)

workingjubilee replied:
Hmm, okay.

workingjubilee reviewed (commented):
This does seem a bit better, but I think we can still keep in mind the role of this repo.

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,22 @@
+  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software" src="https://www.rust-lang.org/static/images/rust-social-wide.jpg" width="512" height="256">
 
-[![Rust Community](https://img.shields.io/badge/Rust_Community%20-Join_us-brightgreen?style=plastic&logo=rust)](https://www.rust-lang.org/community)
+[Website](https://www.rust-lang.org/) | [Getting started](https://www.rust-lang.org/learn/get-started) | [Learn](https://www.rust-lang.org/learn) | [Documentation](https://www.rust-lang.org/learn#learn-use) | [Contributing](CONTRIBUTING.md)
```
That's better!

I think this should either have multiple lines still, separated by the pipe char, or move the links into the "link block" at the end of this section. Whatever you think would be nicer for the inherent split of visualizing the result + diffs, ultimately.

Urgau replied:
Since we are already using "link block" in other parts of the README.md, I decided to use those (and did the same for the tools).

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,22 @@
-# The Rust Programming Language
+<div align="center">
+  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software" src="https://www.rust-lang.org/static/images/rust-social-wide.jpg" width="512" height="256">
```
This doesn't need to all be on one line like it currently is, how about
```suggestion
  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software"
    src="https://www.rust-lang.org/static/images/rust-social-wide.jpg"
    width="512" height="256">
```

Maybe the other way around? :thinking:

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,22 @@
+- **Reliability:** Our rich type system and ownership model ensure memory and thread safety, reducing bugs at compile-time.
 
-</details>
+- **Productivity:** Comprehensive documentation, user-focused compiler, and advanced tooling including package manager and build tool ([Cargo](https://github.com/rust-lang/cargo)), auto-formatter ([rustfmt](https://github.com/rust-lang/rustfmt)), linter ([Clippy](https://github.com/rust-lang/rust-clippy)) and editor support ([rust-analyzer](https://github.com/rust-lang/rust-analyzer)).
```
Hm I think we can express it like...
```suggestion
- **Productivity:** Comprehensive documentation, a compiler committed to providing great diagnostics, and advanced tooling including package manager and build tool ([Cargo](https://github.com/rust-lang/cargo)), auto-formatter ([rustfmt](https://github.com/rust-lang/rustfmt)), linter ([Clippy](https://github.com/rust-lang/rust-clippy)) and editor support ([rust-analyzer](https://github.com/rust-lang/rust-analyzer)).
```

Urgau replied:
Sure, I like this. Done.

workingjubilee reviewed (commented):
Okay cool, this looks much better!

Thank you for figuring out what works best with GitHub, I didn't think it was going to be so peculiar about formatting... :sweat:

matthiaskrgr commented:
The centered logo with background looks kinda weird in gh dark mode imo.

![image](https://github.com/rust-lang/rust/assets/476013/86c1e5b8-6ef0-4eda-9034-347e5ae68ba1)

Urgau commented:
@matthiaskrgr Agree; I tried finding a dark variant of it, but it seems like we don't have one yet.
I'm planning on asking the web-team for a dark-variant that we can integrate.

Unfortunately I'm not sure we can do anything in the mean-time, since we can't use inline styles.

workingjubilee commented:
@Urgau Can you find an svg of the logo? I may have a trick up my sleeve or five.

workingjubilee commented on README.md:
```diff
@@ -1,27 +1,32 @@
-# The Rust Programming Language
+<div align="center">
+  <img alt="The Rust Programming Language: A language empowering everyone to build reliable and efficient software"
+       src="https://www.rust-lang.org/static/images/rust-social-wide.jpg"
+       width="512" height="256">
```
I think we can fix the dark mode thing as a followup maybe but I checked with a couple people and am getting "logo too big" feedback. It's especially annoying on smaller devices, please try the [srcset](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#srcset) attribute as a fix?

workingjubilee replied:
Alternatively twiddling the `width` as a % may be better?

Urgau replied:
What about `width="50%"`?

![Capture d’écran 2024-04-01 à 21 38 15](https://github.com/rust-lang/rust/assets/3616612/fd64e51b-3542-4085-a78f-9cfbabb3a5ec)
![Screenshot 2024-04-01 at 21-38-32 Urgau_rust at improve-readme](https://github.com/rust-lang/rust/assets/3616612/9409d2c9-668e-4d23-a9a8-744ed25bbb2d)

Urgau commented:
> Can you find an svg of the logo? I may have a trick up my sleeve or five.

@workingjubilee It was added to the website in https://github.com/rust-lang/www.rust-lang.org/pull/1003, but unfortunately there wasn't any source provided, nor can't I find one.

But, after some (unfortunately not so) quick editing I came up with this, for a dark variant of it.
![rust-social-wide-dark](https://github.com/rust-lang/rust/assets/3616612/f1b00fbd-3ce2-4f1d-9510-5db1859db2b7)

workingjubilee commented:
Oh nice, this basically seems to look pretty good.

onefifth committed:
Use SVG logos in the README.md.

These should likely eventually reference the rust-lang.org site instead
of raw.githubusercontent.com

Urgau committed:
Add Why Rust? section to the README.md and mention our tools

Urgau commented:
Thanks to @onefifth we now have two beautiful SVG versions of the social banner.

We already updated the PR to include them. Should be ready to merge.

workingjubilee commented:
Nice!

@bors r+

bors commented:
:pushpin: Commit ce4cc9df960ec5c0a840d75e7e57284bfd80d25d has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee ce4cc9df960ec5c0a840d75e7e57284bfd80d25d -->
<!-- homu: {"type":"Approved","sha":"ce4cc9df960ec5c0a840d75e7e57284bfd80d25d","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jhpratt commented:
For some reason, bors still has this in the queue even though it's been merged.

@bors r-
